### ISSUE TITLE
Usage of underscores in aliases

### ISSUE THREAD

**Issue by yvonnefroehlich** [GenericMappingTools/pygmt#2014]
In the Contributors Guide in the section [Code Style](https://www.pygmt.org/dev/contributing.html#code-style):

> When wrapping a new alias, use an underscore to separate words bridged by vowels (aeiou), such as no_skip and z_only. Do not use an underscore to separate words bridged only by consonants, such as distcalc, and crossprofile. This convention is not applied by the code checking tools, but the PyGMT maintainers will comment on any pull requests as needed.

However, some aliases seem not to follow this convention. For example:
- [`pygmt.Figure.plot()`](https://www.pygmt.org/dev/api/generated/pygmt.Figure.plot.html):  `straight_line` (`A`) , `error_bar` (`E`), `nodata` (`d`)
 Following the convention, they should be `straigthline`, `errorbar` , `no_data`
- [`pygmt.Figure.rose()`](https://www.pygmt.org/dev/api/generated/pygmt.Figure.rose.html): `vector_params` (`M`)
 Should be `vectorparams`.

_Please note_: This is primarily meant as a general question for understanding, rather than a request for renaming these aliases. :wink:

**Comment by maxrjones** [comment]
I think that the aliases that do not follow this convention were added to PyGMT before the convention was established. For context, the idea of a convention to ensure consistency for alias names was first suggested in https://github.com/GenericMappingTools/pygmt/pull/1190#issuecomment-814451969 and added in https://github.com/GenericMappingTools/pygmt/pull/1256.

**Comment by yvonnefroehlich** [comment]
Thanks @maxrjones for the explanation and the background information!

**Comment by seisman** [comment]
I don't think we will rename these aliases. So closing the issue.

**Comment by seisman** [comment]
Here is a Python script written by ChatGPT to print all parameters of PyGMT functions and Figure's plotting methods:
```python
import inspect
import pygmt

def list_pygmt_parameters():
    results = {}

    # A helper to extract parameters from a callable
    def extract(name, obj):
        try:
            sig = inspect.signature(obj)
            params = list(sig.parameters.keys())
            results[name] = params
        except (TypeError, ValueError):
            pass

    # 1. Inspect top-level PyGMT module
    for name, obj in inspect.getmembers(pygmt):
        if inspect.isfunction(obj):
            extract(f"pygmt.{name}", obj)

    # 2. Inspect Figure class methods
    for name, obj in inspect.getmembers(pygmt.Figure):
        if inspect.isfunction(obj) or inspect.ismethod(obj):
            # Skip private or dunder methods
            if not name.startswith("_"):
                extract(f"pygmt.Figure.{name}", obj)

    # 3. Inspect submodules (datasets, helpers, etc.)
    for submodule_name, submodule in inspect.getmembers(pygmt, inspect.ismodule):
        prefix = f"pygmt.{submodule_name}"
        for name, obj in inspect.getmembers(submodule):
            if inspect.isfunction(obj):
                extract(f"{prefix}.{name}", obj)

    return results


all_params = list_pygmt_parameters()

# Print each method/function and its parameters
for name, params in list(all_params.items()):
    print(name, params)

# Print all parameters
unique_params = []
for name, params in list(all_params.items()):
    unique_params.extend(params)
for param in sorted(set(unique_params)):
    print(param)
```

The full list of parameters are pasted below:

<details><summary>Details</summary>

```
above
alabel
aliases
alpha
angle
annotate
annotation
anti_alias
anti_aliasing
area_thresh
args
argstr
aspatial
autolabel
azimuth
background
barwidth
bb_style
below
between
binary
bitcolor
blabel
borders
bordervalues
box
categorical
center
circ_subregion
clabel
clearance
close
cmap
coe
color_model
coltypes
combitable
compass
component
compressionfill
confdict
connection
constantfill
continuous
contour
contourpen
convention
convergence
conversions
cores
critical
crop
crossprofile
crs
cstyle
cumulative
cut
cyclic
data
data_source
dcw
deprecate_version
depth
dfile
diameter
direction
discontinuity
dist2line
dist2pt
distance
distcalc
distribution
distribution_name
download
dpi
drapegrid
duplicate
ellipse
empty
encoding
end
endpoint
equalsize
error_bar
event_name
extend
extensionfill
extreme
facadepen
figsize
file
filename_or_obj
fill
fillnegative
fillpositive
filter
filter_type
find
fixedlabel
flat_earth
fmt
fmtfile
fname
font
force
force_scan
frame
func
gap
generate
geographic
geojson
grid
gridfill
gridmask
gs_option
gs_path
gs_version
header
histtype
hole
horizontal
image
imagefile
img_in
incols
infile
inquire
intensity
interpolation
inverse
justify
kind
kwargs
kwdict
label
label_placement
labelbox
labels
lakes
land
latitude
length
levels
limit
line
log
longitude
lonlat
lower
map_scale
mapping
margin
margins
mask
maskvalues
max_retries
maxradius
meshpen
method
mincols
minmax_pos
mode
modname
module_func
monochrome
name
nan_transparent
nans
ncols
ndim
nearest_multiple
neighborfill
newcolname
newname
nlevels
no_bg
no_clip
no_scale
no_skip
no_sort
nodal
nodata
node_table
norm
normalize
nrows
numpoints
offset
oldname
orientation
outcols
outfile
outgrid
outline
output
output_type
override
overrule_bg
panel
pen
per_column
perspective
plane
plot_latitude
plot_longitude
points
pole
polygon
position
prefix
profile
projection
quantile_value
quantity
radiance
radius
region
registration
remove_version
replace
required
resample
rescale
resize
resolution
reverse
rivers
rose
runtimes
scale
scaling
search_radius
sector
sectors
self
sep
separator
series
shading
sharex
sharey
shift
shorelines
show
single_form
size
skip
skiprows
slope_file
sort
source
spacing
spec
speed
splinefill
stack
stairs
statistic
straight_line
style
subsize
suffix
summary
surftype
symbol
tag
tension
terminator
terminator_datetime
text
textfiles
tiles
time_col
timefmt
title
toggle
track
tracks
trackvalues
transparency
transparent
triangular_mesh_pen
truncate
uncertainty
uncertaintyfill
unit
units
upper
use_srtm
use_word
value
vector
vector_params
vectors
verbose
voronoi
wait
waiting
water
weight
width
worldfile
wrap
x
xmax
xmin
xshift
y
ymax
ymin
yshift
z
z_only
z_subregion
zfile
zoom
zoom_adjust
zscale
zsize
zvalue
```

</details>

**Comment by seisman** [comment]
Below is a long list of parameters that contain two words. 

<details><summary>Details</summary>

```
alabel
anti_alias
anti_aliasing
area_thresh
argstr
aspatial
autolabel
barwidth
bb_style
bitcolor
blabel
bordervalues
circ_subregion
clabel
color_model
coltypes
combitable
compressionfill
confdict
constantfill
contourpen
convention
crossprofile
cstyle
data_source
deprecate_version
dfile
dist2line
dist2pt
distance
distcalc
drapegrid
endpoint
equalsize
error_bar
event_name
extensionfill
facadepen
figsize
filename_or_obj
fillnegative
fillpositive
filter_type
fixedlabel
flat_earth
fmtfile
force_scan
gridfill
gridmask
gs_option
gs_path
gs_version
histtype
imagefile
img_in
incols
infile
label_placement
labelbox
lonlat
map_scale
maskvalues
max_retries
maxradius
meshpen
mincols
minmax_pos
module_func
nan_transparent
nearest_multiple
neighborfill
newcolname
newname
nlevels
no_bg
no_clip
no_scale
no_skip
no_sort
nodata
node_table
nrows
numpoints
oldname
outcols
outfile
outgrid
outline
output_type
overrule_bg
per_column
plot_latitude
plot_longitude
quantile_value
remove_version
search_radius
sharex
sharey
single_form
skiprows
slope_file
splinefill
straight_line
subsize
surftype
terminator_datetime
textfiles
time_col
timefmt
trackvalues
triangular_mesh_pen
uncertaintyfill
use_srtm
use_word
vector_params
worldfile
xmax
xmin
xshift
ymax
ymin
yshift
z_only
z_subregion
zfile
zoom_adjust
zscale
zsize
zvalue
```

</details> 

I feel they can be categorized into four groups:

1. **Follow** the current naming convention and **use** underscores, e.g., `anti_alias`, `no_skip`.

<details><summary>Click to see the list</summary>

```
anti_alias
anti_aliasing
area_thresh
data_source
deprecate_version
filename_or_obj
force_scan
img_in
module_func
no_bg
no_clip
no_scale
no_skip
no_sort
node_table
overrule_bg
slope_file
quantile_value
remove_version
single_form
time_col
use_srtm
use_word
```

</details> 

2. **Violate** the current naming convention and **use** underscores, e.g., `color_model`, `event_name`
<details><summary>Click to see the list</summary>

```
bb_style
circ_subregion
color_model
error_bar
event_name
filter_type
flat_earth
gs_option
gs_path
gs_version
label_placement
map_scale
max_retries
minmax_pos
nan_transparent
nearest_multiple
output_type
per_column
plot_latitude
plot_longitude
search_radius
straight_line
terminator_datetime
triangular_mesh_pen
vector_params
z_only
z_subregion
zoom_adjust
```

</details> 

3. **Violate** the current naming convention, **don't use** underscores, and can be more readable with underscores, e.g., `autolabel`

<details><summary>Click to see the list</summary>

```
autolabel
facadepen
splinefill
```

</details>

4. **Follow** the current naming convention, **don't** use underscore, but can be more readable with underscores, e.g., `bordervalues`
<details><summary>Click to see the list</summary>

```
barwidth
bitcolor
bordervalues
combitable
compressionfill
constantfill
contourpen
crossprofile
equalsize
extensionfill
fillnegative
fillpositive
fixedlabel
labelbox
maskvalues
maxradius
meshpen
neighborfill
trackvalues
uncertaintyfill
```

</details> 

5. **Follow** the current naming convention; **don't** use underscore, but is fine without underscores, e.g., `alable`, `xshfit` (some may be controversial)
<details><summary>Click to see the list</summary>

```
alabel
argstr
aspatial
blabel
clabel
coltypes
confdict
cstyle
dfile
dist2line
dist2pt
distance
distcalc
drapegrid
endpoint
figsize
fmtfile
gridfill
gridmask
histtype
imagefile
incols
infile
lonlat
mincols
newcolname
newname
nlevels
nodata
nrows
numpoints
oldname
outcols
outfile
outgrid
outline
sharex
sharey
skiprows
subsize
surftype
textfiles
timefmt
worldfile
xmax
xmin
xshift
ymax
ymin
yshift
zfile
zscale
zsize
zvalue
```

</details> 

I agree with @yvonnefroehlich that always using underscores between two words can make the parameter names more readable, and also avoid some inconsistencies. For example, if we follow our current naming convention strictly, we should use `contourpen`, `meshpen`, and `facade_pen`, which can be quite confusing.

**Comment by seisman** [comment]
In other words, if we decide to follow the current convention strictly, we should fix parameters in group 3 and group 2. If we decide to always use underscores between words (unless the word is already in a short form, e.g., `fmtfile`), we should fix parameters in group 3 and group 4.

**Comment by seisman** [comment]
Ping @GenericMappingTools/pygmt-maintainers for discussions on this issue.

**Comment by yvonnefroehlich** [comment]
Especially from group 2, it's visible that we do not really follow the current convention. I think always using an underscore improves the readability and clarity for users, and for us it's easier to set up new alias names. Thus I would not change the alias names in group 2 to **not** use an underscore. E.g., if we decide to always use an underscore, we could have `spline_fill`, `compression_fill`, `extension_fill`, `negative_fill`, `positive_fill`, `grid_fill` instead of `spline_fill` (currently `splinefill`), `compressionfill`, `extensionfill`, `fillnegative`, `fillpositive`, `gridfill`.

In case we decide to fix the names (in any direction), we should do this soon, as we should definitely use the changed parameter names in the paper. It looks like there are fewer parameters in groups 3 and 4 than in groups 3 and 2. I would like to have the naming more consistent and clearer, and I think it pays off in a long-term view, but I do not know how happy users will be with several parameter name changes...

BTW is there any rule used for the Julia wrapper or for the long names for GMT?

**Comment by weiji14** [comment]
> In case we decide to fix the names (in any direction), we should do this soon, as we should definitely use the changed parameter names in the paper. It looks like there are fewer parameters in groups 3 and 4 than in groups 3 and 2. I would like to have the naming more consistent and clearer, and I think it pays off in a long-term view, but I do not know how happy users will be with several parameter name changes...

Just wanted to float the idea of going for a **breaking** PyGMT v1.0.0 release with all these parameter name changes. Though ideally, we should still go through a deprecation cycle with at least one or two minor releases with the old/new names. So e.g., deprecate in v0.18.0 and v0.19.0, then make jump to v1.0.0.

**Comment by seisman** [comment]
Just to confirm, are you proposing that v1.0.0 be released right after v0.19.0?

**Comment by weiji14** [comment]
Yes, or we could align it with the paper publication (whenever the review is finalized). There are some issues in the [v1.0.0 milestone](https://github.com/GenericMappingTools/pygmt/milestone/11) that could be moved to v2 perhaps if we're serious on a major release next year.

**Comment by seisman** [comment]
I'm a little hesitant about a v1.0.0 release next year, as several parameters still take non-Pythonic arguments.

**Comment by seisman** [comment]
I believe we've reached an agreement to use underscores between words. This also aligns with the [PEP 8 style guide](https://peps.python.org/pep-0008/), which recommends:

> Use the function naming rules: lowercase with words separated by underscores as necessary to improve readability.

That said, we need to rename and deprecate some parameter names, but I feel it may make sense **not** to use underscores for some parameters, especially when a parameter name is already a well-recognized shorthand. For example:

* `surftype` instead of `surf_type` or `surface_type`
* `histtype` instead of `hist_type`
* `timefmt` instead of `time_fmt` or `time_format`

**Comment by yvonnefroehlich** [comment]
> I'm a little hesitant about a v1.0.0 release next year, as several parameters still take non-Pythonic arguments.

Besides this, there is the discussion about dropping the support of the single-letter flags from GMT with the release of v1.0.0. (However, there are some concerns from users.) 

--- 

Do we plan to rename all these aliases for the release of v0.18.0?

**Comment by seisman** [comment]
Yes, I think we should rename most of them, especially those used in the manuscript.

**Comment by seisman** [comment]
@yvonnefroehlich If you want, please also the Contributors Guides.

**Comment by yvonnefroehlich** [comment]
Some more (see also the overview table added at the beginning of this issue based on the script from Dongdong):

* `skiprows`: Would change it to `skip_rows`.
  #4295
* `combitable`: Would change it to `combi_table`.
  #4293
* `runtimes`: Would change it to `run_times`. 
   #4294 
* `margin` for `Figure.inset` but `margins` (plural) for `Figure.suplot`: Would use both times `margins` for **-M**.
  #4296
* `units` for `pygmt.x2sys_init` but `unit` for the rest. 
  #4297
* `startpoint`, `endpoint`, `numpoint` :  Could be fine as is; or `start_point`, `end_point`, `num_point` ?
* `imagefile`, `textfiles`, `worldfile`: Could be fine as is; or `image_file`, `text_files`, `world_file` ?
* `newcolname` but `event_name`, `distribution_name`: Unsure, i.e., `new_col_name` ?

**Comment by seisman** [comment]
> * `margin` for `Figure.inset` but `margins` (plural) for `Figure.suplot`: Would use both times `margins` for **-M**.

Yes.

> * `units` for `pygmt.x2sys_init` but `unit` for the rest.

`x2sys_init` takes a list of units, so `units` is correct.

> * `startpoint`, `endpoint`, `numpoint` :  Could be fine as is; or `start_point`, `end_point`, `num_point` ?

- `startpoint`: I don't find it in the codes.
- `endpoint`: sometimes it's taken as a single word
- `numpoints`: The docstrings say "Give the maximum number of data points on either side of the crossover to use in the spline interpolation [3].". I feel `numpoints` is not a suitable name for it.
> * `imagefile`, `textfiles`, `worldfile`: Could be fine as is; or `image_file`, `text_files`, `world_file` ?

I prefer to keep them unchanged.

> * `newcolname` but `event_name`, `distribution_name`: Unsure, i.e., `new_col_name` ?

I prefer to keep `newcolname`.

**Comment by yvonnefroehlich** [comment]
> `startpoint`: I don't find it in the codes.

Your are right, there is no parameter called `startpoint`. I was mixing it up with `center` for **-C** of [pygmt.project](https://www.pygmt.org/dev/api/generated/pygmt.project.html#pygmt.project).

**Comment by yvonnefroehlich** [comment]
One more (hopefully the last one): We have `max_retries`  for `Figure.tilemap` but `maxradius` for `pygmt.surface`. Leave `maxradius` as is or change it to `max_radius` ?

**Comment by seisman** [comment]
> change it to `max_radius` ?

I think yes.

**Comment by yvonnefroehlich** [comment]
I think we have renamed most of the affected parameter names (see also the overview table in https://github.com/GenericMappingTools/pygmt/issues/2014#issue-1309022433). Some parameter names need more discussions reading the new name or a more Pythonic syntax. `skiprows` and `nodata` remain unchanged as too many functions / methods would be affected.

 * #4275 
 * #4277 
 * #4278
 * #4280 
 * #4293 
 * #4297

**Comment by seisman** [comment]
I guess we can close the issue. We will revisit these PRs later.
